### ISSUE TITLE
Validations performed on missing required params that are members of an optional parameter

### ISSUE THREAD

**Issue by texpert** [ruby-grape/grape#1467]
Like in such a case:
```
params do
  optional :preferences, type: Array do
    requires :key
    requires :value
  end
end
```

Tracked down to this string of code -https://github.com/ruby-grape/grape/blob/master/lib/grape/validations/validators/base.rb#L39

```
  if @required || (resource_params.respond_to?(:key?) && resource_params.key?(attr_name))
   validate_param!(attr_name, resource_params)
  end
```

I think either @required shouldn't be there (if attributes are present, the validation should be performed), or, otherwise, scope.optional parameter should be additionally analyzed.

**Comment by dblock** [comment]
I *think* it makes sense, but  there be dragons. I would start by writing some specs for this.

**Comment by texpert** [comment]
Closing, because it is fixed in Grape 0.17

I wrote the spec, it is failing on Grape 0.16.2 - but discovered that in 0.17 the bug is fixed :)

`      context 'when using requires with a custom validator in an optional block' do
        before do
          subject.resource :items do
            params do
              requires :custom1, customvalidator: true
              optional :option, type: Array do
                requires :custom2, customvalidator: true
              end
            end

            get do
              'optional with custom in optional block works!'
            end
          end
        end

        it "skips validation when parameter isn't present" do
          get '/items', custom1: 'im custom', participants: 'im custom'
          expect(last_response.body).to eq('optional with custom in optional block works!')
        end
      end
`
